### PR TITLE
Draft: Resolve parallel dbt templating issue

### LINKED ISSUE #1088: Dbt templater re-runs compile for each file while running parallel
While debugging slow running using the new `--parallel` feature, I've found that the dbt templater is running the compile step again for each file.

I *think* this is because each file uses it's own `Linter` object and so doesn't utilise the cached methods on the dbt templater.

The solution to this I think is to run the templating step in the main thread before dispatching out for parsing and linting. This should be possible by modifying the runner classes and the method calls available on `Linter`. The tricky part will be splitting out the templating step from `Linter.lint_string()` and `Linter.parse_string()`.

@barrywhart , having thought about this today, I was going to suggest this:
- We *don't* create a new `Linter` for each file.
- Instead, we abstract the parsing and linting methods out into `staticmethods` which can be safely called by the threads. Behind the existing methods of `parse_string()` and `lint_string()` there will be private methods of `_render_string()`, `_parse()` and `_lint()` which are static methods and handle all of the meaningful logic. These can be called via the existing methods _or directly from the runner classes_.
- In the parallel runners, the templating step (`_render_string()`) is called in the main thread, and then the resulting `TemplatedFile` in handed off to a child process to parse and lint (`_parse()` and `_lint()`). Those child processes never need their own `Linter()` object because they're just using the static methods, but might need a copy of the `FluffConfig` or similar objects.

Reckon that could work? I'm happy to have a go at this, I should probably get more familiar with the parallel code as is.

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This PR has a couple of things to still shake out. Namely that one test is failing for a reason I don't understand, and it looks like the CLI output is treading on toes when running in parallel. Those issues aside, I think it's ready for a review, at least from an architecture standpoint.

This resolves #1088 .

I've tested it on our large dbt project with the dbt templater and it's FAST.

It also builds on #1103 so will be easier to review once that's merged.

In short:
- We render the sql files in the main thread and then pass them to the child threads for parsing and linting.
- The rest of the changes are making enough methods in the linter _static_ or _class_ methods to do that in a way that we don't have to create a new `Linter` object each time.

### CODE REVIEW

**Comment by alanmcruickshank** [comment]
The outstanding test is `test__linter__skip_dbt_model_disabled`, which I'm wondering whether this is actually a better solution for. The existing test checks that a `LintedFile` is still returned but that it has a null `templated_file`. In this reworking, no `LintedFile` is returned at all. It's still effectively _skipped_, but skipped in a slightly different way. @NiallRees, @dmateusp  - do you have views on how this might play out for users?

File: src/sqlfluff/core/linter/runner.py [hunk 1]
```diff
@@ -211,3 +207,28 @@ def __init__(self, ee):
...
+
+
+def get_runner(
+    linter,
+    config,
+    parallel: int,
+    allow_process_parallelism: bool = True,
+) -> BaseRunner:
+    """Generate a runner instance based on parallel and sytem configuration."""
+    if parallel > 1 and sys.version_info > (3, 7):
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/runner.py:219]: I'm thinking now that the version check should use `>=`, i.e.: `sys.version_info >= (3, 7)` because it's 3.6.x that is not supporting parallel.

**Comment by barrywhart** [comment]
About the CLI output, could we potentially capture the output in the workers and return it to the main process and print it there? That's what I think I did with the linter output in the first parallel implementation.

File: src/sqlfluff/core/linter/runner.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/linter/runner.py:219]: I've updated the comparison. Do you have something I can link to and put in a comment about *why* 3.6 isn't supported. I had a quick google just now and couldn't work it out.

**Comment by alanmcruickshank** [comment]
> About the CLI output, could we potentially capture the output in the workers and return it to the main process and print it there? That's what I think I did with the linter output in the first parallel implementation.

This is what I _intended_ to do here, but evidently didn't manage. I'll have another look.

File: src/sqlfluff/core/linter/runner.py (hunk 1 above)
Comment (barrywhart) [reply src/sqlfluff/core/linter/runner.py:219]: TBH, I don't remember. I know it wasn't working. I recall thinking we could possibly get it working, but 3.6 is so old, was it worth the effort.

I'll see if I can reproduce the error on 3.6 again.

**Comment by barrywhart** [comment]
@alanmcruickshank: I replied to your question about the issue with 3.6. Mentioning it here in case that comment thread gets hidden by GitHub!!

**Comment by dmateusp** [comment]
@alanmcruickshank I think not returning a LintedFile makes sense!

**Comment by alanmcruickshank** [comment]
I think this is ready to review properly.

The additional changes are:
- Solved the issue with double logging of paths and threaded logging (@barrywhart - I'd just duplicated the code 🤦 - one was inside the thread).
- In debugging this I came across an odd failure with the `fatal` templating errors I added a while back. I've added a counter in the `dbt` templater so that they only actually trigger a fatal error if they happen 3 times in a row. This isn't really related to this PR, but it's a tiny change that was easy to make while I was in the code.
- File names are now explicit fields on `ParsedString` and `RenderedFile`. Previously we introspected these from the contained `TemplatedFile` but in the case of a templating error, this was `None` and so triggered some issues and odd failures.
- There's an additional logging line in the `dbt` templater for the _completion_ of templating so that if you're not running in high verbosity modes it doesn't just look like it's stuck on `templating...` for ages.

**Comment by alanmcruickshank** [comment]
I realise the `--bench` command isn't that useful in a parallel world because the timings overlap. I've added a total `Clock time` output to the timing outputs so that we can evaluate performance better.

On 858 files using the dbt templater:
- `-p 1`: Clock time:   418.02
- `-p 4`: Clock time:   261.33

Not quite a 50% reduction, but I'm still pretty happy with that. 🎉

File: src/sqlfluff/core/linter/linter.py [hunk 2]
```diff
@@ -455,13 +461,16 @@ def lint_parsed(
             # We're only going to return the *initial* errors, rather
             # than any generated during the fixing cycle.
             violations += initial_linting_errors
+        else:
+            # If no parsed tree, set to Nonw
```
Comment (barrywhart) [inline src/sqlfluff/core/linter/linter.py:465]:
```suggestion
            # If no parsed tree, set to None

```

File: src/sqlfluff/core/templaters/dbt.py [hunk 3]
```diff
@@ -199,12 +206,18 @@ def process(self, *, fname, in_str=None, config=None, formatter=None):
...
 
         try:
-            return self._unsafe_process(fname, in_str, config)
+            processed_result = self._unsafe_process(fname, in_str, config)
+            # Reset the fail counter
+            self._sequential_fails = 0
+            return processed_result
         except DbtCompilationException as e:
+            # Increment the counter
+            self._sequential_fails += 1
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/dbt.py:215]: Why would compilation fail the first time but subsequently work?

**Review by barrywhart — APPROVED** [review]
A couple of questions, nothing huge. Looks good!

File: src/sqlfluff/core/templaters/dbt.py (hunk 3 above)
Comment (alanmcruickshank) [reply src/sqlfluff/core/templaters/dbt.py:215]: some compilation fails are fatal (the kind that initially inspired this), but I've realised some aren't. They are where the compilation fails in a file-specific way.

File: src/sqlfluff/cli/commands.py [hunk 4]
```diff
@@ -648,7 +648,7 @@ def parse(path, code_only, format, profiler, bench, nofail, logger=None, **kwarg
                     if parsed
                     else None,
                 )
-                for filepath, (parsed, _, _, _, _) in zip(filepaths, result)
+                for filepath, (parsed, _, _, _, _, _) in zip(filepaths, result)
```
[the remark this answers is not in the document]
Comment (alanmcruickshank) [reply src/sqlfluff/cli/commands.py:657]: Yeah great idea, I think this can be much simpler.

File: src/sqlfluff/cli/commands.py [hunk 5]
```diff
@@ -644,17 +644,16 @@ def parse(path, code_only, format, profiler, bench, nofail, logger=None, **kwarg
...
-                    if parsed
+                    filepath=linted_result.fname,
+                    segments=linted_result.tree.as_record(
+                        code_only=code_only, show_raw=True
+                    )
+                    if linted_result.tree
                     else None,
                 )
-                for filepath, (parsed, _, _, _, _, _) in zip(filepaths, result)
+                for linted_result in result
```
Comment (barrywhart) [inline src/sqlfluff/cli/commands.py:656]: 🎉

**Review by barrywhart — APPROVED** [review]
Very nice!!
